starter/src/test/java/dev/openbanking4/spring/security/multiauth/configurers/collectors/PSD2CollectorTest.java

```diff
@@ -127,9 +133,33 @@
     @Before
     public void setUp() {
 
+        // Build a collector that will not return a username if the cert is not an eidas WEB cert.
         this.psd2Collector = PSD2Collector.psd2Builder()
                 .collectorName("psd2-for-test")
-                .usernameCollector(certificatesChain -> certificatesChain[0].getSubjectDN().getName())
+                .usernameCollector((certificatesChain, psd2CertInfo) -> {
+                    String userName = null;
+                    if(psd2CertInfo != null && psd2CertInfo.isPsd2Cert()){
+                        Optional<EidasCertType> eidasCertType = psd2CertInfo.getEidasCertType();
+                        if (eidasCertType.isPresent()) {
+                            if( eidasCertType.get() == EidasCertType.WEB ) {
+                                if (certificatesChain[0] != null) {
+                                    userName = certificatesChain[0].getSubjectDN().getName();
+                                } else {
+                                    throw new IllegalArgumentException("certificateChain must not contain null " +
+                                            "entries");
+                                }
+                            } else {
+                                log.debug("Expected eidasCertType to be WEB. It was '{}'. Will not collect username",
```

Comment:
There are quite a few indented if statements, which makes it a little hard to read (and determine which else belong to which if). I guess you find the log statements helpful if there are problems, but of course most of the time they will be ignored.

I wonder if it's worth combining them and failing early like this?

```
if (eidasCertType.isEmpty() || eidasCertType.get() != EidasCertType.WEB || certificatesChain[0] == null) {
   fail("Could not extract username from certificate chain");
} else {
   userName = certificatesChain[0].getSubjectDN().getName();
}
```

Reply:
Fixed (ish)

Reply:
I'll take your word for it, but I'm not sure what you've changed! :-)